datafusion/optimizer/src/analyzer.rs

```diff
@@ -172,19 +174,34 @@ fn check_correlations_in_subquery(
         | LogicalPlan::EmptyRelation(_)
         | LogicalPlan::Limit(_)
         | LogicalPlan::Subquery(_)
-        | LogicalPlan::SubqueryAlias(_) => inner_plan.apply_children(|plan| {
-            check_correlations_in_subquery(outer_plan, plan, expr, can_contain_outer_ref)
-        }),
+        | LogicalPlan::SubqueryAlias(_) => {
+            for child in inner_plan.inputs() {
+                child.collect(&mut |plan| {
+                    check_correlations_in_subquery(
+                        outer_plan,
+                        plan,
+                        expr,
+                        can_contain_outer_ref,
+                    )?;
+                    Ok(Recursion::Continue)
+                })?;
+            }
+            Ok(())
+        }
         LogicalPlan::Join(_) => {
             // TODO support correlation columns in the subquery join
-            inner_plan.apply_children(|plan| {
-                check_correlations_in_subquery(
-                    outer_plan,
-                    plan,
-                    expr,
-                    can_contain_outer_ref,
-                )
-            })
+            for child in inner_plan.inputs() {
+                child.collect(&mut |plan| {
+                    check_correlations_in_subquery(
+                        outer_plan,
+                        plan,
+                        expr,
+                        can_contain_outer_ref,
+                    )?;
+                    Ok(Recursion::Continue)
+                })?;
```

Comment:
I think the original `apply_children` just apply the `func` to each child, there is no recursively application.

Reply:
Thanks for pointing this out. Fixed.